pkg/ipamd/ipamd.go

```diff
@@ -1003,66 +1056,55 @@ func (c *IPAMContext) eniIPPoolReconcile(ipPool []string, attachedENI awsutils.E
 				log.Debugf("Reconcile skipping IP %s on ENI %s because it was recently unassigned from the ENI.", strPrivateIPv4, eni)
 				continue
 			} else {
-				log.Debugf("This IP was recently freed, but is out of cooldown. We need to verify with EC2 control plane.")
-				// Call EC2 to verify IPs on this ENI
-				ec2Addresses, err := c.awsClient.GetIPv4sFromEC2(eni)
-				if err != nil {
-					log.Error("Failed to fetch ENI IP addresses!")
-					continue
-				} else {
+				if needEC2Reconcile {
+					// IMDS data might be stale
+					log.Debugf("This IP was recently freed, but is now out of cooldown. We need to verify with EC2 control plane.")
+					// Only call EC2 once for this ENI
+					if ec2VerifiedAddresses == nil {
+						var err error
+						// Call EC2 to verify IPs on this ENI
+						ec2VerifiedAddresses, err = c.awsClient.GetIPv4sFromEC2(eni)
+						if err != nil {
+							log.Errorf("Failed to fetch ENI IP addresses from EC2! %v", err)
+							// Do not delete this IP from the datastore or cooldown until we have confirmed with EC2
+							seenIPs[strPrivateIPv4] = true
+							continue
+						}
+					}
 					// Verify that the IP really belongs to this ENI
 					isReallyAttachedToENI := false
-					for _, ec2Addr := range ec2Addresses {
+					for _, ec2Addr := range ec2VerifiedAddresses {
 						if strPrivateIPv4 == aws.StringValue(ec2Addr.PrivateIpAddress) {
 							isReallyAttachedToENI = true
 							log.Debugf("Verified that IP %s is attached to ENI %s", strPrivateIPv4, eni)
 							break
 						}
 					}
-					if isReallyAttachedToENI {
-						c.reconcileCooldownCache.Remove(strPrivateIPv4)
-					} else {
-						log.Warnf("Skipping IP %s on ENI %s because it does not belong to this ENI!.", strPrivateIPv4, eni)
+					if !isReallyAttachedToENI {
+						log.Warnf("Skipping IP %s on ENI %s because it does not belong to this ENI!", strPrivateIPv4, eni)
 						continue
 					}
 				}
+				// The IP can be removed from the cooldown cache
+				// TODO: Here we could check if the IP is still used by a pod stuck in Terminating state!
```

Comment:
Can we please create an issue to track this? Also to fix we need to query API Server right since kubelet would have deleted it?

Reply:
Is it enough to track through #1091, since that was about pods stuck in terminating?

Reply:
Updated the comment :)